### PR TITLE
List possible and placed supplier orders

### PR DESCRIPTION
**Pull request by fadwamahmoud**
## Summary by CodeRabbit

- **New Features**
	- Introduced a new type for supplier orders, enhancing data structure clarity.
	- Added functionality to retrieve placed supplier orders with detailed information.
	- Implemented new tests to verify supplier order creation and retrieval scenarios.
	- Updated the layout and page components to streamline data handling for supplier orders.

- **Bug Fixes**
	- Improved error handling for scenarios where no supplier orders are found.

- **Documentation**
	- Enhanced type definitions for better clarity and usability.

### CODE REVIEW

**Comment by fadwamahmoud** [comment]
@coderabbitai full review

**Review by delasous — CHANGES_REQUESTED** [review]
As someone looking at these models and methods for the first time, it took me several hours to teeth it apart. Could you please begin to provide context and explanation for the supplier queries, which are all fairly complex and look similar on first glance. It would also be helpful to provide a description for the scope of the PR. I understand now that "listing possible and placed orders" means that the queries that are being wired into the UI are `getPossibleSupplierOrderInfos` and `getPlacedSupplierOrders`, but this was not immediately obvious to me as some of the highlighted changes are to other queries (and I started to get lost reviewing the work that needs to be done on them)
 
- `getPosssibleSupplierOrderInfos` could be changed to `getPossibleSupplierOrders`. Each returned row is a summary of supplier order => it could be ordered by supplier.name or supplier.id, not book.isbn. Where "quantity", "placed" and "price" are referenced, it would be helpful to know which table they reference => `customer_order_lines.quantity|placed` and `book.price`

- dummy data should be removed from `suppliers/+page.svelte`. It is still be referenced in an `#if` block on line 76.

- Unused imports could be removed and `import`, `const` lines could be ordered so that imports come first

File: apps/web-client/src/routes/orders/suppliers/+layout.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { getInitializedDB } from "$lib/db/orders";
+
+import { getPlacedSupplierOrders, getPossibleSupplerOrderInfos } from "$lib/db/orders/suppliers";
```
Comment (silviot) [inline apps/web-client/src/routes/orders/suppliers/+layout.ts:3]: Coderabbit spotted a typo here. I'll apply its suggestions.

File: apps/web-client/src/lib/db/orders/suppliers.ts [hunk 2]
```diff
@@ -50,8 +50,8 @@
 
 export async function getPossibleSupplerOrderLines(db: DB): Promise<SupplierOrderLine[]> {
 	// We need to build a query that will yield all books we can order, grouped by supplier
-	const result = await db.execO<{ supplier_id: number; isbn: string; quantity: number }>(
-		`SELECT supplier_id, book.isbn, SUM(quantity) as quantity
+	const result = await db.execO<SupplierOrderLine>(
+		`SELECT supplier_id, supplier.name as supplier_name, book.isbn, SUM(quantity) as quantity
```
[the remark this answers is not in the document]
Comment (silviot) [reply apps/web-client/src/lib/db/orders/suppliers.ts:54]: The idea is not bad. The diff is `very wrong ®`

File: apps/web-client/src/lib/db/orders/__tests__/supplier-order.test.ts [hunk 3]
```diff
@@ -15,9 +21,9 @@ describe("Suppliers order creation", () => {
 
 	it("sees possible supplier orders from client orders", async () => {
 		expect(await getPossibleSupplerOrderLines(db)).toStrictEqual([
-			{ supplier_id: 1, isbn: "1", quantity: 1 },
-			{ supplier_id: 1, isbn: "2", quantity: 1 },
-			{ supplier_id: 2, isbn: "3", quantity: 2 }
+			{ supplier_id: 1, isbn: "1", quantity: 1, supplier_name: "Science Books LTD" },
+			{ supplier_id: 1, isbn: "2", quantity: 1, supplier_name: "Science Books LTD" },
+			{ supplier_id: 2, isbn: "3", quantity: 2, supplier_name: "Phantasy Books LTD" }
```
[the remark this answers is not in the document]
Comment (silviot) [reply apps/web-client/src/lib/db/orders/__tests__/supplier-order.test.ts:26]: This is true. Tests should be refactored a bit to reduce repetition. But it should be done in a different unit of work.

File: apps/web-client/src/lib/db/orders/suppliers.ts [hunk 4]
```diff
@@ -77,25 +76,28 @@ export async function getPossibleSupplerOrderInfos(db: DB): Promise<SupplierOrde
...
 	const supplierOrderMapping = {};
 	// Collect all supplier ids involved in the order lines
 	const supplierIds = Array.from(new Set(orderLines.map((item) => item.supplier_id)));
+
 	await db.tx(async (passedDb) => {
 		const db: DB = passedDb as DB;
 		for (const supplierId of supplierIds) {
 			// Create a new supplier order for each supplier
-			const newId = (
+			const newSupplierOrderId = (
```
Comment (silviot) [inline apps/web-client/src/lib/db/orders/suppliers.ts:92]: Every time we have a query executed in a loop we should pause and examine if it's really really the only option we have. 99.99999% of the time (100% in my experience, but I consider myself open minded) it's better to add a new (or extend an existing) query to get all the needed data at once.

The idea behind SQL should be that a single query can get you what you need, efficiently.

File: apps/web-client/src/lib/db/orders/suppliers.ts [hunk 5]
```diff
@@ -108,10 +110,10 @@ export async function createSupplierOrder(db: DB, orderLines: SupplierOrderLine[
 
 			let copiesToGo = orderLine.quantity;
 			while (copiesToGo > 0) {
-				const line = customerOrderLines.shift();
-				if (line.quantity <= copiesToGo) {
+				const customerOrderLine = customerOrderLines.shift();
+				if (customerOrderLine) {
```
[the remark this answers is not in the document]
Comment (silviot) [reply apps/web-client/src/lib/db/orders/suppliers.ts:114]: The main problem here is that there are too many queries being performed... `customerOrderLines` being possibly empty is a minor problem.

File: apps/web-client/src/lib/db/orders/__tests__/supplier-order.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (silviot) [reply apps/web-client/src/lib/db/orders/__tests__/supplier-order.test.ts:26]: Yes, please. Do it.

**Review by silviot — CHANGES_REQUESTED** [review]
I think the db queries need some improvements. It's easier for me to take care of it and after the fact articulate the reasons for the change than explaining now how I believe it should be changed. I will work on improving the SQL in `createSupplierOrder`. I'm marking this as "Request changes" to signify it still needs to be worked on.

**Comment by delasous** [comment]
I think updates to the createSupplierOrder query should be handled in a separate issue/pr. This one mainly works with getPossibleSupplierOrders and getSupplierOrders

**Pull request merged**